### PR TITLE
[RLlib|New API|Inconsistency] LSTM Encoder lacks the output Linear, but stated in the docstring (#47625)

### LINKED ISSUE #47625: [RLlib|New API|Inconsistency] LSTM Encoder lacks the output Linear, but stated in the docstring
### **Description:**

There is an inconsistency between the docstring and the implementation of the `TorchLSTMEncoder` class in the RLlib new API. 

### **Current Behavior:**
The docstring of `TorchLSTMEncoder` states that the encoder has "One linear output layer". However, the current implementation does not include this linear output layer.

### **Expected Behavior:**
The implementation should match the docstring, either by:
1. Adding the linear output layer to the implementation, or
2. Removing the mention of the linear output layer from the docstring.

### **Files:**
`ray/rllib/core/models/tf/encoder.py`
`ray/rllib/core/models/torch/encoder.py`

### **Code Snippet:**

```python
class TorchLSTMEncoder(TorchModel, Encoder):
    """A recurrent LSTM encoder.

    This encoder has...
    - Zero or one tokenizers.
    - One or more LSTM layers.
    - One linear output layer.
    """
# ... rest of the implementation
@override(Model)
def _forward(self, inputs: dict, kwargs) -> dict:
    outputs = {}

    # ... existing code ...

    out, states_out = self.lstm(out, (states_in["h"], states_in["c"]))
    states_out = {"h": states_out[0], "c": states_out[1]}

    # !! Note: There's no application of a linear output layer here !!

    # Insert them into the output dict.
    outputs[ENCODER_OUT] = out
    outputs[Columns.STATE_OUT] = tree.map_structure(lambda s: s.transpose(0, 1), states_out)
    return outputs
```

### **Suggested Fix:**
Either:
1. Add a linear output layer in the `__init__` method and use it in the `_forward` method:
   ```python
   self.output_layer = nn.Linear(config.hidden_dim, config.output_dims[0])
   ```
   And in `_forward`:
   ```python
   out = self.output_layer(out)
   ```

2. Remove "- One linear output layer." from the docstring.

### **Additional Context:**
This inconsistency could lead to confusion for developers using or maintaining this encoder. Resolving it will improve code clarity and maintainability. This is the case in GRU and LSTM for Torch and TensorFlow Implementations. Is this what's supposed to happen?

**Action:**
I will be submitting a PR to fix this issue. The PR will be linked to this issue once created.

Labels: bug, documentation, inconsistency

### Versions / Dependencies

Latest Ray Github Version.

Other not relevant, but here it is:
Macos
Python 3.11 (using conda, and docker)

### Reproduction script

Go to the file located in 

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by brieyla1**
I don't know who to add for review

## Why are these changes needed?

Fixes an inconsistency between the docstring and the implementation of the TorchLSTMEncoder class in the RLlib new API.

See Issue for more info

## Related issue number

Closes #47625

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [X] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [x] Release tests (Not sure, ran all tests from python -m pytest -v -s python/ray/tests/test_basic.py like state in https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#testing and didn't run any C++ Tests are none have been changed.)

### CODE REVIEW

**Comment by simonsays1980** [comment]
@brieyla1 Thanks a lot for your PR and the great catch! Because we are rethinking the design of encoders and their configs to make implementation of modules easier for users, we are reluctant to changes in the encoders themselves at this moment and opt for a change of all docstrings.

This PR is great as it is in its code implementation.

**Comment by brieyla1** [comment]
Interesting, I wonder what's the impact on the action processing of directly using an LSTM connected to a head, some of that knowledge may be making the training easier when the VF layers are not shared because of the extra linear processing of the data & their relationships before splitting it.

At the same time, I understand the reluctance to modify the Catalog & Encoder, as models built with it may have issues.. e.g. when reinitializing the RLModule and migrating the weights.

We should do some benchmarks to weigh the benefits of such an "improvement" since we already know the downsides.

**Comment by simonsays1980** [comment]
> Interesting, I wonder what's the impact on the action processing of directly using an LSTM connected to a head, some of that knowledge may be making the training easier when the VF layers are not shared because of the extra linear processing of the data & their relationships before splitting it.
> 
> At the same time, I understand the reluctance to modify the Catalog & Encoder, as models built with it may have issues.. e.g. when reinitializing the RLModule and migrating the weights.
> 
> We should do some benchmarks to weigh the benefits of such an "improvement" since we already know the downsides.

Benchmarking can be done already without adding the linear layer to the encoder but to the head(s): using `post_fcnet_hiddens` adds layers to the head network(s). The more complex your environment is the more layers will be needed to succeed in the task(s). PPO learns Atari Pong with just a single fc-layer after the CNN encoder (should take around 8 minutes on 4 GPUs and 95 `EnvRunner`s). In this case layers are shared between the policy and the critic (just make sure that you tune the `vf_loss_coeff`).
